### DISCUSSION TITLE
Multi-Threading not working

### DISCUSSION THREAD

**Discussion by Hyp5891** [OpenRadioss/OpenRadioss#1848] [Q&A]
Hi all,

I am using the gui job submission tool for windows from [here](https://openradioss.atlassian.net/wiki/spaces/OPENRADIOSS/pages/58785793/Latest+gui+scripts) and LS Prepost (.k-file) for preprocessing.

Multithreading (for example "-nt parameter" set to 4) works/worked without a problem for some models. I am not totally sure yet but I think after adding NRBs (rigid body connections) I cannot run anything else than one thread only (which is obiously slower). When I put higher numbers the solver is not running. Either it stops after the first timestep with no error message or with Message ID44: "Segmentation Violation".

Does anybody else had this issue and knows how to resolve this?

Thanks a lot in advance!

Best regards,
Christian

**Comment by MarianBulla** [comment]
Hi Hyp5891,
thanks for your interest in OpenRadioss :-)
It's hard to figure out the issue. 
Can you add your model, so it can be deeper investigated, please ?

If it is confidential or too large, can you create a smaller one, where this issue also occures ?

Thanks and best regards,
Marian

**Reply by Hyp5891** [reply]
Hi Marian, 

sorry for my late reply.
At the model I am working on right now I observed that contacts seem to play a role in this.
Usually I define *Automatic Surface to Surface for every part interacting with each other.
After building up a certain amount of the model (6, or 7 contacts) multithreading was not working anymore.
I feel that once I defined one part interacting with more then one other part this happend.

Currently I worked around this (in order to drastically reduce runtime 1 thread vs 12 threads!) by defining only one *Automatic Single Surface ("Slave mode").

Did somebody else has discovered this problem?
I will try to build a simple model (like u proposed) soon to study this behavior more deeply.

Thanks in advance!

Best regards,
Christian

**Reply by PaulAltair** [reply]
I've not seen this phenomenon, on any model I've tried, and can't think of anything obvious that would cause it, so it needs further investigation, please do share if you can reproduce on something shareable

**Reply by MarianBulla** [reply]
Hi Christian,
yes, as mentioned by Paul, I also did not see such behavior. 
OpenRadioss is very well suited to handle many parts within one or more contact interfaces. 
Have a look at the full vehicle models, where only one (or very few) contact interfaces are used.
They work very good and robust.
Issues are comming often from materials, which reach their limits. But this can also be solved using a well defined failure criterion.

But a model, which shows this issue, would be helpful to figure out the root cause.

Thanks and best regards,
Marian

**Reply by ktommy2** [reply]
Hi @MarianBulla!
I experience the same problem.
If I calculate with 1 CPU (1 thread), the calculation runs within 88 seconds. If I run it on 8 threads, the calculation is much slower and the calculation time exceeds 300 seconds.

Here you will find a small model with which I can create this problem. I used the same input and carried out the calculation twice.
FYI: i increased the mass scaling because i didn't want to wait too long for the results with 8 threads. I'm aware of the incorrect in the results :P
INPUT:    [input2.zip](https://github.com/OpenRadioss/OpenRadioss/files/14444081/input2.zip)
OUT Files with 1 Thread: [Out_Files_1cpu.zip](https://github.com/OpenRadioss/OpenRadioss/files/14444152/Out_Files_1cpu.zip)
OUT Files with 8 Threads:  [Out_Files_8cpu.zip](https://github.com/OpenRadioss/OpenRadioss/files/14444221/Out_Files_8cpu.zip)

If you agree, could you please convert this topic to an issue?
What I find interesting is that the CPUs are actually under load, but the computing time increases drastically.
I wanted to create an issue, but before I do that, I wanted to see if something similar has already been reported...

FYI, my setup:
Hardware:
HP ZBook 16 G10
	CPU	13th Gen Intel(R) Core(TM) i9-13900HK (x86_64)
	RAM	2x32 GB (5200 MHz)
	GPU	Nvidia RTX 3000 Ada Gen. (Laptop GPU)
	SSD	KXG80ZNV2T04 KIOXIA
	
OS: Windows 10 Pro 22H2

If you need more infos, feel free to contact me!

**Reply by MarianBulla** [reply]
Hi @ktommy2,
yes, I will have a look at you model. I already have an idea, on the root cause for this behavior. But I want to double-check.

I will let you know my findings. 

Thanks and best regards,
Marian

**Reply by OlivierWienholtz** [reply]
Hi Ktommy2

I took a look at your CPU: 
This new architecture has 6 Performance Cores & 8 Efficient cores.
For Radioss computation, I would not use Efficient Cores. So don't go over 6 Threads pls.

Furthermore : on some of those machines, I noticed affinity issues. The default Affinity is wrong.
You can also try to set following environment variable before running your Job:

set KMP_AFFINITY=disabled

Let me know if this works.


--
Olivier

**Reply by ktommy2** [reply]
Hi @OlivierWienholtz, hi @MarianBulla!
Thank you for your asap support.

I know that the uploaded model is small. Originally I had a droptest model with over 250k elements, but for testing/debugging purpose I created a smaller one that reproduces the same problem.

I tried to edit the KMP_Affinity. To start a simulation, i use the executable called openradioss_run_script_for_windows_mpi_ps.
Therefore i modified the batch file (openradioss_run_script_ps.bat) and pasted Olivers code.
I then did a little experiment with different policies. I did one run with each policy (and I closed all my applications during the simulations...)
Here are the results:

Ground zero: 1 thread,                                    ET = 88.3s
KMP_AFFINITY=disabled or none, 6 threads, ET = 365.15s
KMP_AFFINITY=disabled or none, 4 threads, ET = 283.72s
KMP_AFFINITY=disabled or none, 2 threads, ET = 180.18s
KMP_AFFINITY=compact, 6 threads,              ET = 291.15s
KMP_AFFINITY=scatter, 6 threads,                 ET = 257.81s
KMP_AFFINITY=balanced, 6 threads,             ET = 251.83s

**Reply by ktommy2** [reply]
Im definitely not an expert in this topic, but mapping them didnt help either:
set KMP_AFFINITY=explicit, proclist=[0,1,2,3,4,5]

(... i hope, ive done it all right)

**Reply by ktommy2** [reply]
Hi,

i gave another try on my private workstation.
It has an Intel(R) Core(TM) i7-8700K (@ 3.70GHz (x86_64), 3700 MHz,  55301 MB RAM,  75172 MB swap).
I did not set the kmp_affinity for now.

1 thread, ET 131.30 sec
2 threads, ET 94.35 sec
6 threads, ET 83.64 sec
12 threads, ET 106.03 sec (For this small model, this setting was rather counterproductive)

I can therefore confirm that I have the problem on my company notebook and not on my workstation. Is there still a workaround? I will apply for a driver update and report if there is an improvement... damn.

**Reply by ktommy2** [reply]
Hi @MarianBulla, hi @OlivierWienholtz!

a few days have passed and i wanted to ask if you have any other ideas i could try. The update did not bring any speedup.

**Reply by OlivierWienholtz** [reply]
Hello Tommy,
We are currently investigating your issue, in particular trying to run on similar Machine than you.
We keep you in touch.

Best Regards,
Olivier

**Comment by MarianBulla** [comment]
Hi Ktommy2,
yes, I agree with Olivier,
the performance is mostly on the used hardware and the model size. 
Your model is quite small. 
So using more threads or cores does not make much sense/difference.
Anyway, the performance should not drop.
I have an older CPU, without the efficency cores, and I did the benchmark, using your model.
Here ist the result on my machine:
![image](https://github.com/OpenRadioss/OpenRadioss/assets/105412502/496bea5a-ac76-44cd-89c3-8667d6bf6e1d)

My CPU: https://www.intel.com/content/www/us/en/products/sku/213799/intel-core-i711850h-processor-24m-cache-up-to-4-80-ghz/specifications.html

Your CPU: https://www.intel.com/content/www/us/en/products/sku/232144/intel-core-i913900hk-processor-24m-cache-up-to-5-40-ghz/specifications.html

![image](https://github.com/OpenRadioss/OpenRadioss/assets/105412502/20745bb2-8b33-4089-a189-a3a754eec969)


Hope the change of the KMP_AFFINITY will help you. 
Please let us know, how much will bring.

Thanks and best regards,
Marian

**Comment by Hyp5891** [comment]
Hey,

I am thinking about buying a Lenovo P16 Gen2 workstation laptop. Now this recent comments on CPU performance got my attention.
Can anyone rule out that the  Intel® Core™ i7-13700HX or Core™ i7-13850HX can become a problem with OpenRadioss??

Best regards,
Christian

**Reply by MarianBulla** [reply]
Hi Christian,
there is no problem with intel i7 -13700HX CPUs and OpenRadioss. 
It is only a question of performance. 

The newer laptop CPUs are more power efficient.

But the older one still has the AVX512, which provides ~20% more performance compared to AVX2 (see the data sheet ):
![image](https://github.com/OpenRadioss/OpenRadioss/assets/105412502/c046551d-d960-40db-8fb3-dcc375e31d03)

So, if you can get the "older" intel i7 11850H laptop with sufficient memory, it will also do the job very well.

Anyway, both have only 2 memory channels, which will limit the performance.

For more details with the recent intel CPUs, see the response from @OlivierWienholtz, on how to get the best out of it.
The other suggestion would be to go with a workstation instead of a laptop.

Hope it helps you ?

Thansk and best regards,
Marian

**Comment by OlivierWienholtz** [comment]
Hello 

@Ktommy2 

We have done tests on a similar machine than your, it is a 12th Generation I7 with 6 performance (including HyperThreading)+ 8 Efficient cores.
It is similar machine than you only 1 Generation older.
I reproduced your behave regarding the SMP Performance. 2 or more threads are quite slower than 1 on your Model.
I observe same ratio than you.

Interestingly, an MPI run with 2 MPI processes give 1.3 Speedup, 
4 MPI processes give 1.15 from 1 to 4, slightly worse than 2 MPI.

I tried a bigger model with 1 Million finite Elements. 
https://openradioss.atlassian.net/wiki/spaces/OPENRADIOSS/pages/47546369/HPC+Benchmark+Models#1M-Element-Neon-model
I have expected speedup for such architectures:
1.8x from 1 to 2 
1.6x from 2 to 4

12th & 13th Generation Laptop CPUs shows this behave on your small model,
The older 11th Generation Laptop CPUs (Rocket Lake) gives expected SMP performance.

Process Placement on such architecture has also importance: Indeed using Efficient cores slows down the job...

To know : You can guess what are the performance & Efficient cores  using the cpuinfo tool shipped with IntelMPI (you have it when installing OneAPI compiler + IntelMPI):
Performance cores have Threads attached, not the Efficient cores. The "processors" with thread ID=0 from 0-11 are the 6 Performance cores.

![p_e_cores_laptop](https://github.com/OpenRadioss/OpenRadioss/assets/98882516/d4fb1fd0-f1d3-49c1-b4da-a2abef811032)

You can use SMP Pinning to stick your commutation to a core : 
I used OMP_PLACES variables, They are equivalent to KMP_ Variables.
If you set 
OMP_PLACES=0,2

This will stick your SMP threads on logical cpuid=0 & 2 which are 2 P-cores 
You will observe performance slowdown if you run in Mono on core 12 to 19.

Hope it helps,
Olivier W.

**Reply by OlivierWienholtz** [reply]
We continued investigating : 
You may try to set following environment variable on your test system : 

set OMP_WAIT_POLICY=ACTIVE

On the machine I am testing, I could achieve 
1 Thread : 90 s
2 Threads : 67s
4 Threads : 60s 
6 Threads : 63s
The model remains small it is to expect that I won't scale but there is no degradation.

**Reply by ktommy2** [reply]
Hello, sorry for the late reply. I was busy and got sick last week....
The suggested modification to the batch file seems to work well. Thank you very much for that!
Next week I will also test it with a larger model and give you a feedback. After that, we can close the topic and I accept your answer as a solution to this problem!
Thank you for your commitment!
